common/blobstore/s3store/client.go

```diff
@@ -0,0 +1,311 @@
+	"go.uber.org/multierr"
+	"io/ioutil"
+	"math"
+	"net/url"
+)
+
+var (
+	// ErrCheckBucketExists could not verify that bucket directory exists
+	ErrCheckBucketExists = &shared.BadRequestError{Message: "could not verify that bucket directory exists"}
+	// ErrWriteFile could not write file
+	ErrWriteFile = &shared.BadRequestError{Message: "could not write file"}
+	// ErrReadFile could not read file
+	ErrReadFile = &shared.BadRequestError{Message: "could not read file"}
+	// ErrCheckFileExists could not check if file exists
+	ErrCheckFileExists = &shared.BadRequestError{Message: "could not check if file exists"}
+	// ErrDeleteFile could not delete file
+	ErrDeleteFile = &shared.BadRequestError{Message: "could not delete file"}
+	// ErrListFiles could not list files
+	ErrListFiles = &shared.BadRequestError{Message: "could not list files"}
+	// ErrConstructKey could not construct key
+	ErrConstructKey = &shared.BadRequestError{Message: "could not construct key"}
+	// ErrBucketConfigDeserialization bucket config could not be deserialized
+	ErrBucketConfigDeserialization = &shared.BadRequestError{Message: "bucket config could not be deserialized"}
+)
+
+type client struct {
+	s3cli s3iface.S3API
+}
+
+// NewClient returns a new Client backed by file system
+func NewClient(cfg *Config) (blobstore.Client, error) {
+	if err := cfg.Validate(); err != nil {
+		return nil, err
+	}
+
+	sess := session.Must(session.NewSession(&aws.Config{
+		Region: aws.String(cfg.Region),
+	}))
+
+	s3cli := s3.New(sess)
+	return &client{
+		s3cli: s3cli,
+	}, nil
+}
+func s3upload(ctx context.Context, s3api s3iface.S3API, bucket string, key string, b []byte, tags url.Values) error {
+
+	_, err := s3api.PutObjectWithContext(ctx, &s3.PutObjectInput{
+		Bucket:  aws.String(bucket),
+		Key:     aws.String(key),
+		Body:    bytes.NewReader(b),
+		Tagging: aws.String(tags.Encode()),
+	})
+
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+
+			if aerr.Code() == s3.ErrCodeNoSuchBucket {
+				return blobstore.ErrBucketNotExists
+			}
+
+		}
+		return err
+	}
+	return nil
+}
+
+func s3download(ctx context.Context, s3api s3iface.S3API, bucket string, key string) ([]byte, error) {
+
+	result, err := s3api.GetObjectWithContext(ctx, &s3.GetObjectInput{
+		Bucket: aws.String(bucket),
+		Key:    aws.String(key),
+	})
+
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+
+			if aerr.Code() == s3.ErrCodeNoSuchBucket {
+				return nil, blobstore.ErrBucketNotExists
+			}
+
+			if aerr.Code() == s3.ErrCodeNoSuchKey {
+				return nil, ErrReadFile
+			}
+
+		}
+		return nil, err
+
+	}
+
+	defer func() {
+		ierr := result.Body.Close()
+		if ierr != nil {
+			err = multierr.Append(err, ierr)
+		}
+	}()
+
+	body, err := ioutil.ReadAll(result.Body)
+
+	return body, err
+}
+
+func s3gettags(ctx context.Context, s3api s3iface.S3API, bucket string, key string) (map[string]string, error) {
+
+	result, err := s3api.GetObjectTaggingWithContext(ctx, &s3.GetObjectTaggingInput{
```

Comment:
Do you need to close anything on this result resource like you do for download?

Reply:
AFAIK there is no need to do this for GetObjectTagging as it does not involve any streams